<?xml version="1.0"?>
<sdf version="1.7">
  <model name="Schunk_Gripper">
    <!-- This model is based on the Schunk WSG 050-110-P gripper
    (https://schunk.com/fileadmin/pim/docs/IM0026091.PDF) with TRI's Punyo
    bubble fingers attached (https://punyo.tech/#instructions). However, the
    bubbles themselves are not specfied file. Instead, they are set up
    programmatically as deformable geometries in the application file.
    Axes:
      +Z - Pointing away from the fingertips.
      +X - Pointing towards the camera.
    Origin:
      (0, 0, 0) at the bottom center of the gripper body. -->
    <pose>0 0 0.06 0 1.5708 0</pose>
    <link name="gripper">
      <pose>0 0 0 0 0 0</pose>
      <inertial>
        <mass>0.988882</mass>
        <inertia>
          <ixx>0.162992</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.162992</iyy>
          <iyz>0</iyz>
          <izz>0.164814</izz>
        </inertia>
        <pose>0 0 0 0 0 0</pose>
      </inertial>
      <visual name="gripper mesh">
        <pose>0 0 -0.036 -1.5708 0 -1.5708</pose>
        <geometry>
          <mesh>
            <uri>package://drake_models/wsg_50_description/meshes/wsg_body.obj</uri>
            <scale>1 1 1</scale>
          </mesh>
        </geometry>
        <material>
          <ambient>0 0 0 0</ambient>
          <diffuse>0.525 0.525 0.525 1</diffuse>
          <specular>0.525 0.525 0.525 1</specular>
        </material>
      </visual>
    </link>
    <link name="left_finger">
      <pose>0 0 -0.0725 0 0 0</pose>
      <inertial>
        <!-- These inertial properties are bogus, but that's ok because we use
         a very stiff PD-controller to dictate the motion of the gripper anyway
        -->
        <mass>0.05</mass>
        <inertia>
          <ixx>0.16</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.16</iyy>
          <iyz>0</iyz>
          <izz>0.16</izz>
        </inertia>
        <pose>0 0 0 0 0 0</pose>
      </inertial>
      <visual name="left_finger_visual_mesh">
        <pose>0 0.0 0 1.5708 0 3.14159</pose>
        <geometry>
          <mesh>
            <uri>package://drake_models/wsg_50_description/meshes/bubble_finger.obj</uri>
            <scale>.001 .001 .001</scale>
          </mesh>
        </geometry>
      </visual>
    </link>
    <link name="right_finger">
      <pose>0 0 -0.0725 0 0 0</pose>
      <inertial>
        <!-- These inertial properties are bogus, but that's ok because we use
         a very stiff PD-controller to dictate the motion of the gripper anyway
        -->
        <mass>0.05</mass>
        <inertia>
          <ixx>0.16</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.16</iyy>
          <iyz>0</iyz>
          <izz>0.16</izz>
        </inertia>
        <pose>0 0 0 0 0 0</pose>
      </inertial>
      <visual name="right_finger_visual_mesh">
        <pose>0 0 0 1.5708 0 0</pose>
        <geometry>
          <mesh>
            <uri>package://drake_models/wsg_50_description/meshes/bubble_finger.obj</uri>
            <scale>.001 .001 .001</scale>
          </mesh>
        </geometry>
        <!-- Avoid rendering the finger so it doesn't block the bubble -->
        <drake:accepting_renderer>nothing</drake:accepting_renderer>"
      </visual>
    </link>
    <joint name="translate_joint" type="prismatic">
      <parent>world</parent>
      <child>gripper</child>
      <axis>
        <xyz expressed_in="__model__">-1 0 0</xyz>
      </axis>
      <drake:controller_gains p='10000.0' d='1.0' />
    </joint>
    <joint name="left_finger_sliding_joint" type="prismatic">
      <parent>gripper</parent>
      <child>left_finger</child>
      <pose>0 0 0 0 0 0</pose>
      <axis>
        <xyz>0 1 0</xyz>
        <limit>
          <lower>-0.055</lower>
          <upper>0</upper>
        </limit>
      </axis>
      <drake:controller_gains p='10000.0' d='1.0' />
    </joint>
    <joint name="right_finger_sliding_joint" type="prismatic">
      <parent>gripper</parent>
      <child>right_finger</child>
      <drake:mimic joint='left_finger_sliding_joint' multiplier='-1' offset='0.0'/>
      <pose>0 0 0 0 0 0</pose>
      <axis>
        <xyz>0 1 0</xyz>
        <limit>
          <lower>0</lower>
          <upper>0.055</upper>
           <!-- Drake attaches an actuator to all joints if the effort limit isn't
           explicitly set to zero. We do NOT want an actuator for this joint due
           to the existence of the mimic tag. -->
          <effort>0</effort>
        </limit>
      </axis>
    </joint>
  </model>
</sdf>